File: C2021/dso7.c
// Copyright (c) 1980, InfoCom Computers and Communications, Cambridge MA 02142
// All rights reserved, commercial usage strictly prohibited.
// Written by R. M. Supnik.
// Revisions Copyright (c) 2021, Darth Spectra (Lydia Marie Williamson).
#include "extern.h"
#include "common.h"

// Encrypt password
void encryp(const char *inw, char *outw/*, size_t inw_unit, size_t outw_unit*/) {
// Initialized data
   const char keyw[1 * 6] = "E" "C" "O" "R" "M" "S";

   int i, j, uinw[6], usum;
   char ukeyw[1 * 6];
   int uinws, ukeyws;

// Parameter adjustments
   --outw;
   --inw;

// Function Body

   uinws = 0;
// 						!UNBIASED INW SUM.
   ukeyws = 0;
// 						!UNBIASED KEYW SUM.
   j = 1;
// 						!POINTER IN KEYWORD.
   for (i = 1; i <= 6; ++i) {
// 						!UNBIAS, COMPUTE SUMS.
      ukeyw[i - 1] = keyw[i - 1] - 64;
      if (inw[j] <= '@') {
         j = 1;
      }
//    uinw[i - 1] = (char)(inw[j]) - 64)
      uinw[i - 1] = inw[j] - 64;
      ukeyws += ukeyw[i - 1];
      uinws += uinw[i - 1];
      ++j;
// L100:
   }

   usum = uinws % 8 + (ukeyws % 8 << 3);
// 						!COMPUTE MASK.
   for (i = 1; i <= 6; ++i) {
      j = (uinw[i - 1] ^ ukeyw[i - 1] ^ usum) & 31;
      usum = (usum + 1) % 32;
      if (j > 26) {
         j %= 26;
      }
      outw[i] = max(1, j) + 64;
// L200:
   }
}

// Move to next state in puzzle room
void cpgoto(int st) {
// System generated locals
   int i__1, i__2;

// Local variables
   int i;

// CPGOTO, PAGE 2

   rooms.rflag[CPuzzRX - 1] &= ~SeenR;
   i__1 = objcts.olnt;
   for (i = 1; i <= i__1; ++i) {
// 						!RELOCATE OBJECTS.
      if (objcts.oroom[i - 1] == CPuzzRX && (objcts.oflag2[i - 1] & ActrO + VillO) == 0) {
         i__2 = findex.cphere * hfactr;
         newsta(i, 0, i__2, 0, 0);
      }
      if (objcts.oroom[i - 1] == st * hfactr) {
         newsta(i, 0, CPuzzRX, 0, 0);
      }
// L100:
   }
   findex.cphere = st;
}

// Describe puzzle room
void cpinfo(int rmk, int st) {
// Local variables
   int i, j, k, l;
   char dgm[1 * 8];

// Initialized data
   static const int dgmoft[8] = { -9, -8, -7, -1, 1, 7, 8, 9 };
   static const char pict[1 * 5] = "S" "S" "S" " " "M";
   static const char qmk[1] = "?";

// CPINFO, PAGE 2

   rspeak(rmk);
   for (i = 1; i <= 8; ++i) {
      j = dgmoft[i - 1];
      dgm[i - 1] = pict[puzzle.cpvec[st + j - 1] + 3];
// 						!GET PICTURE ELEMENT.
      if (abs(j) == 1 || abs(j) == 8) {
         goto L100;
      }
      k = 8;
      if (j < 0) {
         k = -8;
      }
// 						!GET ORTHO DIR.
      l = j - k;
      if (puzzle.cpvec[st + k - 1] != 0 && puzzle.cpvec[st + l - 1] != 0) {
         dgm[i - 1] = qmk[0];
      }
   L100:
      ;
   }
// write(outch, //F
//    "       |%A2%1X%A2%1X%A2|%/" //F
//    " West  |%A2 .. %A2|  East%/" //F
//    "       |%A2%1X%A2%1X%A2|", //F
//    dgm //F
// ); //F
   more_output("       |%c%c %c%c %c%c|\n", dgm[0], dgm[0], dgm[1], dgm[1], dgm[2], dgm[2]);
   more_output(" West  |%c%c .. %c%c|  East\n", dgm[3], dgm[3], dgm[4], dgm[4]);
   more_output("       |%c%c %c%c %c%c|\n", dgm[5], dgm[5], dgm[6], dgm[6], dgm[7], dgm[7]);

   if (st == 10) {
      rspeak(870);
   }
// 						!AT HOLE?
   if (st == 37) {
      rspeak(871);
   }
// 						!AT NICHE?
   i = 872;
// 						!DOOR OPEN?
   if (findex.cpoutf) {
      i = 873;
   }
   if (st == 52) {
      rspeak(i);
   }
// 						!AT DOOR?
   if (puzzle.cpvec[st] == -2) {
      rspeak(874);
   }
// 						!EAST LADDER?
   if (puzzle.cpvec[st - 2] == -3) {
      rspeak(875);
   }
// 						!WEST LADDER?
}

File: dso7.f
C Copyright (c) 1980, InfoCom Computers and Communications, Cambridge MA 02142
C All rights reserved, commercial usage strictly prohibited.
C Written by R. M. Supnik.
C Revisions Copyright (c) 2021, Darth Spectra (Lydia Marie Williamson).
C
C Encrypt password
	SUBROUTINE ENCRYP(INW,OUTW)
	IMPLICIT INTEGER(A-Z)
	CHARACTER INW(6),OUTW(6)
	CHARACTER  KEYW(6),UKEYW(6)
	INTEGER UINW(6)
	DATA KEYW/'E','C','O','R','M','S'/
C
	UINWS=0
C						!UNBIASED INW SUM.
	UKEYWS=0
C						!UNBIASED KEYW SUM.
	J=1
C						!POINTER IN KEYWORD.
	DO 100 I=1,6
C						!UNBIAS, COMPUTE SUMS.
	  UKEYW(I)=char(ichar(KEYW(I))-64)
	  IF(INW(J).LE.char(64)) J=1
C	  UINW(I)=char(ichar(INW(J))-64)
	  UINW(I)=ichar(INW(J))-64
	  UKEYWS=UKEYWS+ichar(UKEYW(I))
	  UINWS=UINWS+UINW(I)
	  J=J+1
100	CONTINUE
C
	USUM=MOD(UINWS,8)+(8*MOD(UKEYWS,8))
C						!COMPUTE MASK.
	DO 200 I=1,6
	  J=IAND(IEOR(IEOR(UINW(I),ichar(UKEYW(I))),USUM),31)
	  USUM=MOD(USUM+1,32)
	  IF(J.GT.26) J=MOD(J,26)
	  OUTW(I)=char(MAX0(1,J)+64)
200	CONTINUE
	RETURN
	END
C
C Move to next state in puzzle room
	SUBROUTINE CPGOTO(ST)
	IMPLICIT INTEGER(A-Z)
C
	COMMON /HYPER/ HFACTR
	include 'rooms.h'
	include 'rflag.h'
	include 'rindex.h'
	include 'objects.h'
	include 'oflags.h'
	include 'flags.h'
C CPGOTO, PAGE 2
C
	RFLAG(CPUZZ)=IAND(RFLAG(CPUZZ),not(RSEEN))
	DO 100 I=1,OLNT
C						!RELOCATE OBJECTS.
	  IF((OROOM(I).EQ.CPUZZ).AND.
     &	(IAND(OFLAG2(I),(ACTRBT+VILLBT)).EQ.0))
     &	CALL NEWSTA(I,0,CPHERE*HFACTR,0,0)
	  IF(OROOM(I).EQ.(ST*HFACTR))
     &	CALL NEWSTA(I,0,CPUZZ,0,0)
100	CONTINUE
	CPHERE=ST
	RETURN
	END
C
C Describe puzzle room
	SUBROUTINE CPINFO(RMK,ST)
	IMPLICIT INTEGER(A-Z)
	INTEGER DGMOFT(8)
	CHARACTER  DGM(8),PICT(5)
C
	COMMON /CHAN/ INPCH,OUTCH,INDEXCH,STORYCH
C
C PUZZLE ROOM
C
	COMMON /PUZZLE/ CPDR(16),CPWL(8),CPVEC(64)
	include 'flags.h'
C
C FUNCTIONS AND LOCAL DATA
C
C
	DATA DGMOFT/-9,-8,-7,-1,1,7,8,9/
	DATA PICT/'S','S','S',' ','M'/
C CPINFO, PAGE 2
C
	CALL RSPEAK(RMK)
	DO 100 I=1,8
	  J=DGMOFT(I)
	  DGM(I)=PICT(CPVEC(ST+J)+4)
C						!GET PICTURE ELEMENT.
	  IF((IABS(J).EQ.1).OR.(IABS(J).EQ.8)) GO TO 100
	  K=8
	  IF(J.LT.0) K=-8
C						!GET ORTHO DIR.
	  L=J-K
	  IF((CPVEC(ST+K).NE.0).AND.(CPVEC(ST+L).NE.0))
     &	DGM(I)='?'
100	CONTINUE
 	WRITE(OUTCH,10) DGM
C
	IF(ST.EQ.10) CALL RSPEAK(870)
C						!AT HOLE?
	IF(ST.EQ.37) CALL RSPEAK(871)
C						!AT NICHE?
	I=872
C						!DOOR OPEN?
	IF(CPOUTF) I=873
	IF(ST.EQ.52) CALL RSPEAK(I)
C						!AT DOOR?
	IF(CPVEC(ST+1).EQ.-2) CALL RSPEAK(874)
C						!EAST LADDER?
	IF(CPVEC(ST-1).EQ.-3) CALL RSPEAK(875)
C						!WEST LADDER?
	RETURN
C
10	FORMAT('       |',A2,1X,A2,1X,A2,'|'/,
     &' West  |',A2,' .. ',A2,'|  East',/
     &'       |',A2,1X,A2,1X,A2,'|')
	END
